File: src/com/pokosho/util/StringUtils.java
package com.pokosho.util;
import com.pokosho.db.Pos;

public class StringUtils {
	public static String ENCODE_STRING = "UTF-8";
	public static final int KUROMOJI_POS_INDEX = 0;
	private final static String NOUN = "名詞";
	private final static String VERV = "動詞";
	private final static String ADJECTIVE = "形容詞";
	private final static String ADVERB = "副詞";
	private final static String PRONOUN = "代名詞";
	private final static String PREPOSITION = "前置詞";
	private final static String CONJUNCTION = "接続詞";
	private final static String INTERJECTION = "感動詞";
	private final static String RENTAI = "連体詞";
	private final static String JOSHI = "助詞";
	private final static String UNKNOWN = "未知語"; /* 名詞扱い */

	// private final static String OTHER = "記号";

	/**
	 * 文字列をシンプルにする. URLを除いたり、全角半角統一、「」削除を行う.
	 * 
	 * @param str
	 * @return
	 */
	public static String simplize(String str) {
		// URLを削除
		String res = str
				.replaceAll(
						"^(https?|ftp)(:\\/\\/[-_.!~*\\'()a-zA-Z0-9;\\/?:\\@&=+\\$,%#]+)",
						"");
		res = res.replaceAll("@[a-zA-Z0-9_]+", "");
		res = res.replaceAll("[「【(（『].*[」】）』¥¥(¥¥)¥¥[¥¥]]", "");
		res = res.replaceAll("\"", "");
		res = res.replaceAll("[\uE000-\uF8FF]", "。"); // 携帯の絵文字
		res = res.replaceAll("[\u1F004-\u1F6C0]", "。"); // unicodeの絵文字
		return res;
	}

	public static String simplizeForReply(String str) {
		String res = StringUtils.simplize(str);
		res = res.replaceAll("[¥¥!！¥¥?？¥¥.。¥¥-]", "");
		return res;
	}

	public static Pos toPos(String posStr) {
		String p = posStr.split("-")[0];
		// log.debug("posStr:" + posStr + " toPos:" + p + " detail:" + posStr);
		if (p == null || p.length() == 0 || p.startsWith(UNKNOWN)) {
			/* 未知語は名詞扱い */
			return Pos.Noun;
		}
		if (p.startsWith(NOUN)) {
			return Pos.Noun;
		}
		if (p.startsWith(VERV)) {
			return Pos.Verv;
		}
		if (p.startsWith(INTERJECTION)) {
			return Pos.Interjection;
		}
		if (p.startsWith(ADJECTIVE)) {
			return Pos.Adjective;
		}
		if (p.startsWith(ADVERB)) {
			return Pos.Adverb;
		}
		if (p.startsWith(PRONOUN)) {
			return Pos.Preposition;
		}
		if (p.startsWith(CONJUNCTION)) {
			return Pos.Conjunction;
		}
		if (p.startsWith(PREPOSITION)) {
			return Pos.Preposition;
		}
		if (p.startsWith(RENTAI)) {
			return Pos.Rentai;
		}
		if (p.startsWith(JOSHI)) {
			return Pos.Joshi;
		}
		return Pos.Other;
	}
}

File: src/main/kotlin/com/pokosho/util/StringUtils.kt
package com.pokosho.util

import com.pokosho.db.Pos
import java.util.regex.Pattern

object StringUtils {
    var ENCODE_STRING = "UTF-8"
    private val ALFABET_PATTERN = Pattern.compile(
        "[a-zA-Z:\\-]", Pattern.CASE_INSENSITIVE
    )
    private val CONTAIN_JPN_PATTERN = Pattern.compile(
        "[ぁ-んァ-ヴ一-龠]+", Pattern.CASE_INSENSITIVE
    )

    private val HANGUL_PATTERN = Pattern.compile(
        "[\uAC00-\uD79F]", Pattern.CASE_INSENSITIVE
    )
    private val NOUN = "名詞"
    private val VERV = "動詞"
    private val ADJECTIVE = "形容詞"
    private val ADVERB = "副詞"
    private val PRONOUN = "代名詞"
    private val PREPOSITION = "前置詞"
    private val CONJUNCTION = "接続詞"
    private val INTERJECTION = "感動詞"
    private val RENTAI = "連体詞"
    private val JOSHI = "助詞"
    private val UNKNOWN = "未知語" /* 名詞扱い */

    /**
     * 文字列をシンプルにする. URLを除いたり、全角半角統一、「」削除を行う.
     *
     * @param str
     * @return
     */
    fun simplify(str: String): String {
        // URLを削除
        var res = str
            .replace("^(https?|ftp)(:\\/\\/[-_.!~*\\'()a-zA-Z0-9;\\/?:\\@&=+\\$,%#]+)".toRegex(), "")
        res = res.replace("@[a-zA-Z0-9_]+".toRegex(), "")
        res = res.replace("[「【(（『].*[」】）』¥¥(¥¥)¥¥[¥¥]]".toRegex(), "")
        res = res.replace("\"".toRegex(), "")
        res = res.replace("[\uE000-\uF8FF]".toRegex(), "。") // 携帯の絵文字
        return res
    }

    fun simplifyForReply(str: String): String {
        return StringUtils.simplify(str).replace("[¥¥!！¥¥?？¥¥.。¥¥-]".toRegex(), "")
    }

    fun toPos(posStr: String): Pos {
        val p = posStr.split("-".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()[0]
        // log.debug("posStr:" + posStr + " toPos:" + p + " detail:" + posStr);
        if (p.length == 0 || p.startsWith(UNKNOWN)) {
            /* 未知語は名詞扱い */
            return Pos.Noun
        }
        if (p.startsWith(NOUN)) {
            return Pos.Noun
        }
        if (p.startsWith(VERV)) {
            return Pos.Verv
        }
        if (p.startsWith(INTERJECTION)) {
            return Pos.Interjection
        }
        if (p.startsWith(ADJECTIVE)) {
            return Pos.Adjective
        }
        if (p.startsWith(ADVERB)) {
            return Pos.Adverb
        }
        if (p.startsWith(PRONOUN)) {
            return Pos.Preposition
        }
        if (p.startsWith(CONJUNCTION)) {
            return Pos.Conjunction
        }
        if (p.startsWith(PREPOSITION)) {
            return Pos.Preposition
        }
        if (p.startsWith(RENTAI)) {
            return Pos.Rentai
        }
        return if (p.startsWith(JOSHI)) {
            Pos.Joshi
        } else Pos.Other
    }

    fun isAlfabet(c: Char): Boolean {
        val matcher = ALFABET_PATTERN.matcher(Character.toString(c))
        return matcher.find()
    }

    fun containsJPN(tweet: String): Boolean {
        val matcher = CONTAIN_JPN_PATTERN.matcher(tweet)
        return matcher.find()
    }

    fun containsKR(tweet: String): Boolean {
        val matcher = HANGUL_PATTERN.matcher(tweet)
        return matcher.find()
    }

    fun containsSurrogatePair(tweet: String): Boolean {
        for (c in tweet.toCharArray()) {
            if (Character.isLowSurrogate(c) || Character.isHighSurrogate(c)) {
                return true
            }
        }
        return false
    }
}
